File: benchmarks/sdk-compare/cpp/src/core/schedule.cpp
#include "schedule.h"

#include <limits>
#include <stdexcept>
#include <string_view>

namespace sdk_compare {
namespace {

BenchmarkSchedule::TimePoint add_seconds(BenchmarkSchedule::TimePoint instant,
                                         std::uint64_t seconds,
                                         std::string_view description) {
    using Clock = BenchmarkSchedule::Clock;
    using Seconds = std::chrono::seconds;

    if (seconds >
        static_cast<std::uint64_t>(std::numeric_limits<Seconds::rep>::max())) {
        throw std::runtime_error(std::string(description) + " overflowed");
    }

    const auto duration =
        std::chrono::duration_cast<Clock::duration>(Seconds(seconds));
    if (duration > Clock::time_point::max() - instant) {
        throw std::runtime_error(std::string(description) + " overflowed");
    }
    return instant + duration;
}

}  // namespace

BenchmarkSchedule BenchmarkSchedule::start(const Execution& execution) {
    const auto origin = Clock::now();
    const auto measurement_start =
        add_seconds(origin, execution.warmup_seconds, "warm-up deadline");
    const auto measurement_end = add_seconds(measurement_start,
                                             execution.measurement_seconds,
                                             "measurement deadline");
    const auto completion_deadline =
        add_seconds(measurement_end,
                    execution.drain_timeout_seconds,
                    "measurement drain deadline");
    return {origin, measurement_start, measurement_end, completion_deadline};
}

BenchmarkSchedule::BenchmarkSchedule(TimePoint origin,
                                     TimePoint measurement_start,
                                     TimePoint measurement_end,
                                     TimePoint completion_deadline)
    : origin_(origin)
    , measurement_start_(measurement_start)
    , measurement_end_(measurement_end)
    , completion_deadline_(completion_deadline) {}

bool BenchmarkSchedule::is_measurement_instant(TimePoint instant) const {
    return instant >= measurement_start_ && instant < measurement_end_;
}

std::uint64_t BenchmarkSchedule::ns_at(TimePoint instant) const {
    if (instant < origin_) {
        throw std::runtime_error("benchmark instant is before schedule origin");
    }
    const auto elapsed =
        std::chrono::duration_cast<std::chrono::nanoseconds>(instant - origin_)
            .count();
    if (elapsed < 0) {
        throw std::runtime_error(
            "benchmark schedule produced a negative timestamp");
    }
    return static_cast<std::uint64_t>(elapsed);
}

std::uint64_t BenchmarkSchedule::now_ns() const { return ns_at(Clock::now()); }

BenchmarkSchedule::TimePoint BenchmarkSchedule::measurement_start() const {
    return measurement_start_;
}

BenchmarkSchedule::TimePoint BenchmarkSchedule::measurement_end() const {
    return measurement_end_;
}

BenchmarkSchedule::TimePoint BenchmarkSchedule::completion_deadline() const {
    return completion_deadline_;
}

TDuration remaining_timeout(BenchmarkSchedule::TimePoint deadline) {
    const auto now = BenchmarkSchedule::Clock::now();
    if (now >= deadline) {
        return TDuration::Zero();
    }
    const auto micros =
        std::chrono::duration_cast<std::chrono::microseconds>(deadline - now)
            .count();
    return TDuration::MicroSeconds(static_cast<std::uint64_t>(micros));
}

}  // namespace sdk_compare

File: ydb/benches/sdk_compare/schedule.rs
use std::time::{Duration, Instant};

use anyhow::{Context, Result};

#[derive(Clone, Copy)]
pub(crate) struct BenchmarkSchedule {
    origin: Instant,
    pub(crate) measurement_start: Instant,
    pub(crate) measurement_end: Instant,
    pub(crate) completion_deadline: Instant,
}

impl BenchmarkSchedule {
    pub(crate) fn new(
        warmup_duration: Duration,
        measurement_duration: Duration,
        drain_timeout: Duration,
    ) -> Result<Self> {
        let origin = Instant::now();
        let measurement_start = origin
            .checked_add(warmup_duration)
            .context("warm-up deadline overflowed")?;
        let measurement_end = measurement_start
            .checked_add(measurement_duration)
            .context("measurement deadline overflowed")?;
        let completion_deadline = measurement_end
            .checked_add(drain_timeout)
            .context("measurement drain deadline overflowed")?;

        Ok(Self {
            origin,
            measurement_start,
            measurement_end,
            completion_deadline,
        })
    }

    pub(crate) fn is_measurement_instant(&self, instant: Instant) -> bool {
        instant >= self.measurement_start && instant < self.measurement_end
    }

    pub(crate) fn ns_at(&self, instant: Instant) -> Result<u64> {
        let elapsed = instant
            .checked_duration_since(self.origin)
            .context("benchmark instant is before schedule origin")?;
        u64::try_from(elapsed.as_nanos())
            .context("benchmark schedule does not fit into u64 nanoseconds")
    }

    pub(crate) fn now_ns(&self) -> Result<u64> {
        self.ns_at(Instant::now())
    }
}
